src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -714,21 +687,44 @@ uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
         return UCS_ERR_NO_DEVICE;
     }
 
-    uct_cuda_copy_sync_memops(md, address, is_vmm);
+    if (cuda_mem_ctx == NULL) {
+        status = uct_cuda_copy_md_primary_ctx_retain(cuda_device,
+                                                     &cuda_mem_ctx);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        primary_ctx_retained = 1;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxPushCurrent(cuda_mem_ctx));
+    if (status != UCS_OK) {
+        if (primary_ctx_retained) {
+            UCT_CUDADRV_FUNC_LOG_WARN(cuDevicePrimaryCtxRelease(cuda_device));
+        }
+
+        return status;
+    }
+
+    uct_cuda_copy_sync_memops(address, is_vmm);
```

Comment:
this code seems over complicated, i would refactor it as follows:

```c
uct_cuda_copy_get_address_range(CUcontext cuda_mem_ctx, CUdevice cuda_device,...)
{
   CUcontext primary_context;

   if (cuda_mem_ctx == NULL) {
      cuDevicePrimaryCtxRetain(&primary_context)
      cuCtxPushCurrent(primary_context)
   }

   cuMemGetAddressRange(...)

   if (cuda_mem_ctx == NULL) {
      cuCtxPopCurrent()
     cuDevicePrimaryCtxRelease(cuda_device)
   }
}
```

```c
if (md->config.alloc_whole_reg == UCS_CONFIG_OFF) {
   goto out_default_range;
}

uct_cuda_copy_get_address_range(cuda_mem_ctx, cuda_device, ....)

```

Reply:
but `cuda_mem_ctx` can be non NULL but not set, then we have to push it also

Reply:
i think cuda_mem_ctx is initialized to NULL

Reply:
anyway the point is to simplify the code and maybe use a helper function to wrap addr range detection with optional primary context push

Reply:
Updated the code including `uct_cuda_copy_sync_memops` into the code wrapped by push and pop context methods.